### ISSUE TITLE
Say all not working in Google Docs

### ISSUE THREAD

**Issue by Qchristensen** [nvaccess/nvda#8210]
In Google Docs on Chrome and Firefox, when reading a document, reading and navigation keys do not work in browse mode, and in focus mode only navigation works, Say all does not work.  Screen reader support is enabled.  Log files created with NVDA Next 15060.4709bfa7, however the results appear to be the same with NVDA 2018.1.1.

In Chrome, in Browse mode:

No error is generated but NVDA reads "about:blank" instead of the text when navigating with the arrows:
```
IO - inputCore.InputManager.executeGesture (11:43:16.043):
Input: kb(desktop):upArrow
IO - speech.speak (11:43:16.062):
Speaking [LangChangeCommand ('en_GB'), u'about:blank']
```
When using Say All, there is no error and again NVDA reads "about:blank", using the Say All profile:
```
IO - inputCore.InputManager.executeGesture (11:43:20.667):
Input: kb(desktop):NVDA+downArrow
DEBUG - queueHandler.registerGeneratorObject (11:43:20.677):
Adding generator 136
IO - speech.speak (11:43:20.709):
Speaking [IndexCommand(1), LangChangeCommand ('en_GB'), u'about:blank']
IO - speech.speak (11:43:20.740):
Speaking [IndexCommand(2)]
DEBUG - queueHandler.pumpAll (11:43:21.711):
generator 136 finished
```

In focus mode, a warning is generated, but the line is read when moving with the arrow keys:
```
IO - inputCore.InputManager.executeGesture (11:40:36.226):
Input: kb(desktop):upArrow
DEBUGWARNING - NVDAObjects.IAccessible.IAccessible._get_IAccessibleRole (11:40:36.243):
accRole failed: (-2147024809, 'The parameter is incorrect.', (None, None, None, 0, None))
DEBUGWARNING - NVDAObjects.IAccessible.IAccessible._get_IAccessibleRole (11:40:36.246):
accRole failed: (-2147024809, 'The parameter is incorrect.', (None, None, None, 0, None))
IO - speech.speak (11:40:36.256):
Speaking [LangChangeCommand ('en_GB'), u'Bullet Conference call for topics ']
```
However nothing is read, and no error is generated when pressing Say All:
```
IO - inputCore.InputManager.executeGesture (11:40:39.634):
Input: kb(desktop):NVDA+downArrow
DEBUG - queueHandler.registerGeneratorObject (11:40:39.650):
Adding generator 44
IO - speech.speak (11:40:39.663):
Speaking [IndexCommand(1)]
DEBUG - queueHandler.pumpAll (11:40:40.163):
generator 44 finished
```
In Firefox, in Browse mode, navigating reads "about:blank" and Say All reads "About:Blank" in the Say all profile.  In Focus mode, warnings are generated, but the text is read (including other editors) when navigating and nothing is read for Say All:

Navigating:
```
IO - inputCore.InputManager.executeGesture (12:10:27.973):
Input: kb(desktop):upArrow
IO - speech.speak (12:10:28.010):
Speaking [LangChangeCommand ('en_GB'), u'about:blank']
```
Say all:
```
IO - inputCore.InputManager.executeGesture (12:10:30.621):
Input: kb(desktop):NVDA+downArrow
DEBUG - queueHandler.registerGeneratorObject (12:10:30.631):
Adding generator 320
IO - speech.speak (12:10:30.665):
Speaking [IndexCommand(1), LangChangeCommand ('en_GB'), u'about:blank']
IO - speech.speak (12:10:30.678):
Speaking [IndexCommand(2)]
IO - speech.speak (12:10:31.194):
Speaking [IndexCommand(3)]
DEBUG - queueHandler.pumpAll (12:10:31.697):
generator 320 finished
```
In focus mode the text is read but a warning generated when navigating:
```
IO - inputCore.InputManager.executeGesture (12:12:04.750):
Input: kb(desktop):upArrow
DEBUGWARNING - NVDAObjects.IAccessible.IAccessible._get_IAccessibleRole (12:12:04.775):
accRole failed: (-2147024809, 'The parameter is incorrect.', (None, None, None, 0, None))
DEBUGWARNING - NVDAObjects.IAccessible.IAccessible._get_IAccessibleRole (12:12:04.779):
accRole failed: (-2147024809, 'The parameter is incorrect.', (None, None, None, 0, None))
DEBUGWARNING - NVDAObjects.IAccessible.IAccessible._get_IAccessibleRole (12:12:04.783):
accRole failed: (-2147024809, 'The parameter is incorrect.', (None, None, None, 0, None))
IO - speech.speak (12:12:04.786):
Speaking [LangChangeCommand ('en_GB'), u'Quentin Christensen is editing this list item. Bullet Conference call for topics  ']
DEBUGWARNING - NVDAObjects.IAccessible.IAccessible._get_IAccessibleRole (12:12:06.243):
accRole failed: (-2147024809, 'The parameter is incorrect.', (None, None, None, 0, None))
```
When using Say All, no warning is generated and nothing at all is read:
```
IO - inputCore.InputManager.executeGesture (12:12:12.910):
Input: kb(desktop):NVDA+downArrow
DEBUG - queueHandler.registerGeneratorObject (12:12:12.911):
Adding generator 345
IO - speech.speak (12:12:12.927):
Speaking [IndexCommand(1)]
DEBUG - queueHandler.pumpAll (12:12:13.428):
generator 345 finished
IO - inputCore.InputManager.executeGesture (12:12:15.742):
```

**Comment by Brian1Gaff** [comment]
Is this a lack of spotting the changes issue like when you sometimes find a 
new page is not read till you bring up a menu and cancel it again?

Brian

**Comment by rperez030** [comment]
I don't think so. I tried bringing up the menu, canceling and retrying the command, and it didn't seem to make any difference.

**Comment by leonardder** [comment]
@Qchristensen: Have you tried this with braille support enabled? When braille support is disabled, say all not working is actually expected.

**Comment by Qchristensen** [comment]
Is it deliberate?  (As in is it a known bug, or specifically supposed to be like that for.... .reasons?)  Screen reader users who don't have a Braille display aren't necessarily going to think to try that option.

I just tried with NVDA 2019.1.1 and Chrome Version 74.0.3729.157 (Official Build) (64-bit), as well as In Firefox 66.0.5 (64-bit) (the results were the same in both browsers):

WIth Screen reader support enabled and Braille support disabled:
In Browse mode, arrows read "blank".  In focus mode, they move and NVDA reads.
In either browse or focus mode, "Say all" reads nothing (NVDA does not say anything).

With Braille support enabled:
arrows and say all both work in either browse or focus mode.

**Comment by Adriani90** [comment]
@Qchristensen I think this needs to be solved by Google, I don't think NVDA can do anything about it since they seem to decide what's exposed and split them in these two cathegories (screen reader suport) and (braille support). In my view, I think screen reader support is kind of a support for people who need a screen reader but still have some vision while braille support is for people which are completely blind. At least I think Google cathegorizes in this way the two settings.
@leonardder what do you think? If you guys agree, I think we can close this issue and may address the cathegorization "screen reader support" and "braille support" directly with Google.

**Comment by feerrenrut** [comment]
I haven't looked into this issue. But if it is true that it requires work from Google we should flag it with them. I wouldn't want this closed until we confirm the cause of the issue.

**Comment by Adriani90** [comment]
Maybe @aleventhal or @ObjectInSpace can give some more details here, Why is there a difference between screenreader support and braille support at all in Google Docs? I am asking because for example Say all feature does not make the difference in NVDA between speech and braille.
Could screenreader support and braille support not be just merged to one setting which improves accessibility in general?

**Comment by ObjectInSpace** [comment]
Thanks for the feedback ;)

The keyboard shortcut for say all with braille support off is ctrl+alt+r. (If memory serves; it's in the help article) This however does not move the focus of the cursor, so it is not ideal for large documents. The screenreader say all features should work as expected when braille mode is on.

**Comment by Adriani90** [comment]
Since this works when braille support is on, I am closing this issue. However, with braille support in Google Docs enabled, it seems the performance is somehow lower. So Hopefully there will be some improvements on this from Google's side.

**Comment by bhavyashah** [comment]
@ObjectInSpace  Could you please explain why say all works only when Braille support is enabled? Also, why does Google Docs have a native say all key command (Ctrl+Alt+R) when all screen readers already offer it themselves? It may be the case that there is little that can be done from NVDA's end for this ticket, but I do not think the problem has been adequately addressed in and of itself. At best, this ticket merits the NeedsExternalFix label. I don't think, however, that WorksForMe correctly describes the status of this issue at all.

**Comment by ObjectInSpace** [comment]
Docs is not my PA.

**Comment by aleventhal** [comment]
Hi @bhavyashah, as @ObjectInSpace mentioned, what is shown on the screen is displayed in a way that isn't accessible to screen readers. In order to enable screen reader support, there are 2 accessibility modes:
- In non-Braille mode, a live region used for announcements. Because there is no position within a normal editor DOM, a screen readers "say all" feature would not work.
- In Braille mode, a hidden fake editor is provided, which is different from what is shown visibly. The screen reader's position is within the hidden fake editor, which contains most of the document's content. The screen reader's "say all" feature should mostly work here, with the caveat that in a long document, some of the content is not loaded until it is scrolled to. Therefore, even in Braille mode, Docs needs to provide its own "say all" in order to truly read everything.
